File: SieveOfEratosthenes/C/src/Sieve.c
#include "../include/Sieve.h"

int * Sieve(int n) {
  // Create an array to hold the sieve
  int * sieve = malloc(n*sizeof(int));

  // Initialize the array with all natural numbers
  for (size_t i = 0; i < n; i++) {
    sieve[i] = 1;
  }

  for(int i = 2; i < sqrt(n); i++) {
    if(sieve[i] == 1) {
      for (size_t j = i*i; j < n; j+=i) {
        sieve[j] = 0;
      }
    }
  }
  printPrimes(sieve, n);
  return sieve;
}

void printPrimes(int * sieve, int size) {
  printf("\nPrimes in range %d are:\n\t", size);
  for (int i = 2; i < size; i++) {
    if(sieve[i] == 1) {
      printf(" %d", i);
    }
  }
}

File: SieveOfEratosthenes/Java/Sieve.java
import java.util.*;

public class Sieve {
  private int size, sieve[], primeCount;
  Sieve(int n) {
    this.size = n;
    sieve = new int[n];
    initializeWithZeros();
    sievify();
  }

  void initializeWithZeros() {
    for(int i = 2; i < this.size; i++) {
      this.sieve[i] = 1;
    }
  }

  void sievify() {
    this.primeCount = this.size - 2; // Neglect 0 and 1

    for(int i = 0; i < Math.sqrt(this.size); i++) {
      if(this.sieve[i] == 1) { 
        for(int j = i*i; j < this.size; j += i) {
          if(this.sieve[j] == 1) {
            this.primeCount--;
          }
          this.sieve[j] = 0;
        }
      }
    }

  }

  int getPrimeCount() {
    return this.primeCount;
  }

  void printPrimes() {
    int counter = 1;
    System.out.println("-------------------");
    for(int i = 0; i < this.size; i++) {
      if(this.sieve[i] == 1) {

        System.out.print(i + " ");

        counter++;

        if(counter % 5 == 0) {
          System.out.println();
        }
      }
    }
    System.out.println("\n-------------------");
  }

  public static void main(String a[])  {
    int size;
    Scanner sc = new Scanner(System.in);
    System.out.println("Enter the size of the sieve: ");
    size = sc.nextInt();
    sc.close();
    
    Sieve soe = new Sieve(size);
    soe.printPrimes();
    System.out.println("There are " + soe.getPrimeCount() + " prime numbers in the sieve.");
  }
}
